### PR TITLE
LPS-115144 Improve performance of the sign-in modal

### PR DESCRIPTION
**Pull request by markocikos**
@liferay-frontend 

Currently, sign in modal is implemented with this order of execution:
1. A user clicks on the 'sign in' link. We start `fetch` to get sign in modal content
1. Once fetch returns content, we open modal.

This causes lag between click and opening of modal, affecting perceived performance.

In this PR, we are changing the behavior, by adding prefetching and loading indicator:
1. A user hovers or focuses on the 'sign in' link. We start `fetch` to get sign in modal content. I would estimate this buys us half a second or more. Focus listener is there to maintain improvements with keyboard-based navigation.
1. If the `fetch` completes before modal is opened, we store response in the `html` variable. When modal opens, we show content immediately. This is also true if the modal is closed and opened again.
1. If the modal is opened while `fetch` is still in progress, we show modal with the loading indicator. When the `fetch` completes, we update modal content. Note that loading time will always be reduced, by the time between hover and click.

Notes:
- I simulated the lag by wrapping `fetch` with `setTimeout`
- There is an edge case where `fetch` may complete during modal opening. [This line](https://github.com/liferay-frontend/liferay-portal/commit/1d9992eb99cd56b0d212919703f77967745f2cc4#diff-56e2a84b2bb09df6cca77f21f9ba7e2aR149) safeguards that case.
- I moved the modal init from frontend-theme to product navigation. I was already considering doing this in https://github.com/wincent/liferay-portal/pull/208 , but what stopped me was alleged usage in [unstyled theme](https://github.com/liferay/liferay-portal/blob/296ff93c5fcb45f4659fa3049cc3c74d7dcb6414/modules/apps/frontend-theme/frontend-theme-unstyled/src/main/resources/META-INF/resources/_unstyled/templates/portal_normal.ftl#L40). However, I made a goof there, the template has `sign-in` **id** and not the class I was looking for. So, we should be safe to move the code. After the move, we have an additional performance benefit - listener init DOM query will not be executed on every single page.
- By default, the content determines modal height. To prevent height change glitch when the login page loads, we need to set height in advance. To achieve this, I extended `openModal` API with `height`, set the height to 400px, and modified styles to centrally align spinner. Alternative would be to change modal to full-screen, but I didn't like the way it looked.
- I considered always fetching eagerly on page load, but that would add unnecessary requests in logged-out navigation.
- I tried rendering `login-web` inside jsp, using `<liferay-portlet:runtime>`, and then reading it from DOM and passing to modal. This approach had some issues, like JS in `login-web` not executing correctly and styles being off as the tag does not support exclusive mode. Issues may be worked around, but it would be a bit ugly solution. Also, we would be adding a direct dependency instead of generic login API point `/c/portal/login`, which may be a bad practice.

### CODE REVIEW

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
Nice @markocikos! Could you maybe record a small gif comparing the before and after, just for show? :)

Maybe even do a "poor-man" measurement using the console and measuring the difference between click and iframe load or sth like that. Just so that we have an idea of the numbers here, nothing too fancy :)

**Comment by wincent** [comment]
Nice idea. It makes sense to try this out here in a one-off special case, but we should also consider whether this is the kind of thing we want to reify into a pattern, because there are plenty of other buttons-that-show-modals in liferay-portal, and I imagine some of them would be ripe for this kind of treatment.

I have a couple of general comments here and one or two things I want to comment on inline.

> 2\. If the `fetch` completes before modal is opened, we store response in the `html` variable. When modal opens, we show content immediately. This is also true if the modal is closed and opened again.

The closed-and-opened again bit worries me a bit, although maybe I am just being paranoid. The modal contains a form, right? And that form is going to have some kind of potentially time-sensitive CRSF token embedded inside it. I'm worried that if we cache this and show it again, there is a chance that the form submission won't work on the second render, because the CRSF token will be stale.

Sadly, this is all speculation, because my portal instance is broken right now and I can't even inspect the sign-in modal; I'm just assuming that it has a token in it — maybe it doesn't: in my broken form (which for some reason is not rendering as a modal) I see just this, which is evidently just a timestamp.

```
<input class="field form-control" id="_com_liferay_login_web_portlet_LoginPortlet_formDate" name="_com_liferay_login_web_portlet_LoginPortlet_formDate" type="hidden" value="1592321827285">
```

What do you think?

> * I simulated the lag by wrapping `fetch` with `setTimeout`

Pro-Tip™: This is also pretty handy for that:

Screenshot 2020-06-16 -173109-wdVozOop@2x

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 1]
```diff
@@ -57,5 +57,103 @@
 		<span class="sign-in text-default" role="presentation">
 			<aui:icon cssClass="sign-in text-default" data="<%= anchorData %>" image="user" label="sign-in" markupView="lexicon" url="<%= themeDisplay.getURLSignIn() %>" />
 		</span>
+
+		<aui:script>
+			(function () {
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:62]: You don't need to use an IIFE pattern inside of `aui:script` tags (it already does that for you).

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 2]
```diff
@@ -57,5 +57,103 @@
...
+				var signInLink = document.querySelector('.sign-in > a');
+
+				if (!signInLink || signInLink.dataset.redirect !== 'false') {
+					return;
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:66]: Although, obviously, if you want an early `return`, you do. Just be aware that the client is going to get this code:

```js
(function () {
    (function () {
        // stuff in here
    })();
})();
```

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 3]
```diff
@@ -57,5 +57,103 @@
...
+
+				var signInURL = '<%= themeDisplay.getURLSignIn() %>';
+
+				var modalSignInURL = Liferay.Util.addParams(
+					'windowState=exclusive',
+					signInURL
+				);
+
+				var setModalContent = function (html) {
+					var fragment = document.createRange().createContextualFragment(html);
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:77]: May as well move this into the `if`, otherwise the code looks like this:

```js
const thing = prepareSomethingThatMightNotBeUsed();

if (checkSomething()) {
    useThing(thing);
}
```

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 4]
```diff
@@ -57,5 +57,103 @@
...
+						})
+						.catch(function () {
+							redirect = true;
+						});
+				};
+
+				signInLink.addEventListener('mouseover', function (event) {
+					if (loading || html) {
+						return;
+					}
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:123]: Instead of repeating this `if` in both listeners, you could move it inside `fetchModalSignIn()`.

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp [hunk 5]
```diff
@@ -57,5 +57,103 @@
...
+
+					fetchModalSignIn();
+				});
+
+				signInLink.addEventListener('click', function (event) {
+					event.preventDefault();
+
+					if (redirect) {
+						Liferay.Util.navigate(signInURL);
```
Comment (wincent) [inline modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:140]: Seems we should have a `return` here, right?

**Comment by wincent** [comment]
Assigning back to you @markocikos for your response.

File: modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp (hunk 1 above)
Comment (markocikos) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:62]: That is not always true, it only wraps code in AUI sandbox IIFE if an AUI module is used.

`<aui:script use='aui-base'>` page source:

screen2

`<aui:script>` page source:

screen

By using IIFE in `<aui:script>` we are not polluting global namespace.

Comment (jbalsas) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:62]: That's partially correct, @markocikos!

I think you could/should add `sandbox="<%= true %>"` to `aui:script` to be safe. Could you give that I try? Just talking from memory...

This might be another good usecase for our [POC Props mechanism for React JSP Components](https://github.com/liferay-frontend/liferay-portal/pull/2) to test...

Comment (markocikos) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:62]: Yep, that works as well!

`<aui:script sandbox="<%= true %>">` page source:

screen

Comment (wincent) [reply modules/apps/product-navigation/product-navigation-user-personal-bar-web/src/main/resources/META-INF/resources/view.jsp:62]: Thanks for correcting my misunderstanding. I was basing that on what I had seen in the served code in one example, didn't look at the source.

**Comment by markocikos** [comment]
> The closed-and-opened again bit worries me a bit, although maybe I am just being paranoid. The modal contains a form, right? And that form is going to have some kind of potentially time-sensitive CRSF token embedded inside it. I'm worried that if we cache this and show it again, there is a chance that the form submission won't work on the second render, because the CRSF token will be stale.
> 
> Sadly, this is all speculation, because my portal instance is broken right now and I can't even inspect the sign-in modal; I'm just assuming that it has a token in it — maybe it doesn't: in my broken form (which for some reason is not rendering as a modal) I see just this, which is evidently just a timestamp.

I am not sure if there is a mechanism, a job, that updates our CRSF token in `Liferay.authToken`. If the token does expire on server side, which I'm not sure it does, the `fetch` in current solution would fail as well after some time.

Safeguarding against this is probably overly defensive, maybe even unwanted. If I opened a page, left for an hour and returned, I would consider it reasonable if something security-related fails.

**Comment by markocikos** [comment]
> It makes sense to try this out here in a one-off special case, but we should also consider whether this is the kind of thing we want to reify into a pattern, because there are plenty of other buttons-that-show-modals in liferay-portal, and I imagine some of them would be ripe for this kind of treatment.

It's up to @jbalsas , but I could make an analysis, maybe add couple of implementation suggestions. I guess the first step is to try to measure performance benefit, https://github.com/liferay-frontend/liferay-portal/pull/13#issuecomment-644753470, which I am just about to do.

**Comment by jbalsas** [comment]
> The closed-and-opened again bit [...]

Forgot to comment about this. This approach already showed to be problematic in that a lot of caching is often unwanted. You can grep `destroyOnHide` to get an idea about how many places we purposefully disable this level of caching.

As you can see, modals used to have this by default, but that produces several unexpected scenarios related to obsolete data
- Open/Close a user selecting modal. Add an additional user. Open the user modal again. The user is not there :(
- Open/Close a user selecting modal. Remove the user. Open the user modal again and make an action over the deleted user. Probable NullPointer exception.
- ...

Over years we just had a constant stream of issues that soon surpassed the value of caching the views.

> It's up to @jbalsas , but I could make an analysis, maybe add couple of implementation suggestions. I guess the first step is to try to measure performance benefit, #13 (comment), which I am just about to do.

It's in fact more up to @wincent, he's our tech lead! :) If you agree there's value here and that we should work on it, I can figure out if it's aligned with our priorities and give the green light, but you'll need to pitch it to me :)

Finally, I'm still waiting on my **initial request** 😉 

> Nice @markocikos! Could you maybe record a small gif comparing the before and after, just for show? :)
>
> Maybe even do a "poor-man" measurement using the console and measuring the difference between click and iframe load or sth like that. Just so that we have an idea of the numbers here, nothing too fancy :)

**Comment by wincent** [comment]
> > It's up to @jbalsas , but I could make an analysis, maybe add couple of implementation suggestions. I guess the first step is to try to measure performance benefit, #13 (comment), which I am just about to do.
> 
> It's in fact more up to @wincent, he's our tech lead! :) If you agree there's value here and that we should work on it, I can figure out if it's aligned with our priorities and give the green light, but you'll need to pitch it to me :)

Premature abstraction is the root of all evil. Too often engineers think they've come up with the One Grand Unified Pattern and seek to code it up and then roll it out across the universe, only to find the idea was half-baked. That's why I said:

> It makes sense to try this out here in a one-off special case, but we should also consider whether this is the kind of thing we want to reify into a pattern

Which means: we should do it exactly like you've done here, but *bear in mind* that this might be a pattern we want to turn into an API as we encounter more examples where it would be useful. ie. not yet.

> Over years we just had a constant stream of issues that soon surpassed the value of caching the views.

Ok, you convinced me. Today is not the day to tackle one of [the two hard problems](https://martinfowler.com/bliki/TwoHardThings.html) of computer science.

---

To summarize:

1. Don't bother abstracting this yet.
2. Minor refactor to not cache `html` across close-then-open events.
3. Ship it!

**Comment by markocikos** [comment]
A followup in https://github.com/liferay-frontend/liferay-portal/pull/23
